File: go/560.subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
    return HashMap(nums, k)
}

// TC O(n)
// SC O(n)
func HashMap(nums []int, k int) int {
    // map from 'sum' to '# of occurance of sum'
    sums := map[int]int{
        0: 1,
    }
   
    var sum, count int
    for _, num := range nums {
        sum += num
        if numberOfBegin, exists := sums[sum-k]; exists {
            count += numberOfBegin
        }
        sums[sum]++
    }
    return count
}

// TC O(n^2)
// SC O(n)
// SC can be optimized to O(1)
func NaiveDP(nums []int, k int) int {
    count := 0
    sums := make([]int, len(nums))
    for i, sum := range nums {
        sums[i] = sum
        if sum == k {
            count++
        }
    }
    for length := 2; length <= len(nums); length++ {
        for i := 0; i + length <= len(nums); i++ {
            sums[i] = sums[i] + nums[i+length-1]
            if sums[i] == k {
                count++
            }
        }
    }
    return count
}

File: csharp/560.subarray-sum-equals-k.cs
/*
 * @lc app=leetcode id=560 lang=csharp
 *
 * [560] Subarray Sum Equals K
 */

// @lc code=start

using System.Collections.Generic;

public class Solution 
{
    public int SubarraySum(int[] nums, int k) 
    {
        // return BruteForce(nums, k);
        return PrefixSum(nums, k);
    }

    public int PrefixSum(int[] nums, int k)
    {
        int count = 0;
        int[] prefixSums = new int[nums.Length];

        // prefixSums[i] = sum from nums[0] to nums[i] inclusive.
        prefixSums[0] = nums[0];
        for (int i = 1; i < nums.Length; i++)
        {
            prefixSums[i] = prefixSums[i - 1] + nums[i];
        }

        // (prefix_sum, number_of_occurrence(before current element))
        Dictionary<int, int> prefixSumCounts = new Dictionary<int, int>();
        foreach (int sum in prefixSums)
        {
            if (sum == k)
            {
                count++;
            }

            // add number of occurrences of the sub target prior to current element
            int subTarget = sum - k;
            if (prefixSumCounts.TryGetValue(subTarget, out int prefixSumCount))
            {
                count += prefixSumCount;
            }

            // put current prefix sum into map
            if (prefixSumCounts.TryGetValue(sum, out int c))
            {
                prefixSumCounts[sum] = c + 1;
            }
            else 
            {
                prefixSumCounts[sum] = 1;
            }
        }

        return count;
    }

    // may pass if you have good luck.
    public int BruteForce(int[] nums, int k)
    {
        int count = 0;
        int sum = 0;
        int[] sums = new int[nums.Length];

        for (int i = 0; i < nums.Length; i++)
        {
            for (int j = 0; j + i < nums.Length; j++)
            {
                sum = sums[j] + nums[j + i];
                sums[j] = sum;
                if (sum == k)
                {
                    count++;
                }
            }
        }

        return count;
    }
}
// @lc code=end

